enhancements/cluster-logging/forward_to_cloudwatch.md

```diff
@@ -0,0 +1,198 @@
+
+### User Stories
+
+#### I want to forward logs to Cloudwatch instead of a local store
+
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+spec:
+  outputs:
+   - name: CloudwatchOut
+     type: cloudwatch
+     cloudwatch:
+       region: myregion
+     secret:
+        name: mysecret
+  pipelines:
+  - inputRefs: [application, infrastructure, audit]
+  - outputRefs: [CloudwatchOut]
+```
+
+[Cloudwatch][concepts] defines Log Groups and Log Streams.
+The CRD above creates 3 Log Groups named "application", "infrastructure" and "audit".
+A new Log Stream is created for each container producing logs.
+
+**Note**: The installer UI (Addon or OLM) can get AWS credentials from a `cloudcredential.openshift.io/v1`.
+The user only has to provide a `region` to enable cloudwatch forwarding for a cluster.
+Details are out of scope for this proposal.
+
+#### I want to define my own log groups
+
+There are two ways to define custom log groups:
+- Use a Kubernetes label with the group name.
+- Define the group name in the `ClusterLogForwarder.Output`
+
+1. Using a kubernetes label named `LogGroup`
+
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+spec:
+  outputs:
+   - name: CloudwatchOut
+     type: cloudwatch
+     cloudwatch:
+       region: myregion
+	   groupLabel: LogGroup
+     secret:
+        name: mysecret
+  pipelines:
+  - inputRefs: [application, infrastructure, audit]
+  - outputRefs: [CloudwatchOut]
+```
+
+2. Defining a group in the output spec and use input selectors:
+
+```
+apiVersion: "logging.openshift.io/v1"
+kind: "ClusterLogForwarder"
+  inputs:
+  - name: FooIn
+    application:
+	  namespaces: [foo]
+  - name: BarIn
+    application:
+	  namespaces: [bar]
+  outputs:
+   - name: FooOut
+     type: cloudwatch
+     cloudwatch:
+	   group: foo
+   - name: BarOut
+     type: cloudwatch
+     cloudwatch:
+	   group: bar
+  pipelines:
+  - inputRefs: [FooIn]
+    outputRefs: [FooOut]
+  - inputRefs: [BarIn]
+    outputRefs: [BarOut]
+```
+
+#### I want to use EKS authentication with Cloudwatch
+
+TODO: is this a requirement? If so define an appropriate set of `secret` keys.
+
+### Implementation Details
+
+Using the [fluentd cloudwatch plugin][plugin].
+Most of the CRD fields have an obvious corresponding fluentd parameter.
+
+Set `auto_create_stream true` to create streams and groups on the fly.
+
+`log-stream-name` is set to the `containerID` of the log record.
+This matches the intended use of logs streams, from the [Cloudwatch documentation][concepts]:
+
+> More specifically, a log stream is generally intended to represent the sequence of events coming from the application instance or resource being monitored.
+> For example, a log stream may be associated with an Apache access log on a specific host.
+
+To group logs from multiple containers (by namespace, label etc.) use *Log Groups* as shown in the User Stories section above.
```

Comment:
Seems like there are two default modes that would be fairly easy to consider, and then the one advanced mode for groups:

1. Three main groups: "application", "infrastructure", "audit"
2. Group by "infrastructure", "audit", then by "namespace" for application logs
3. Give the user the ability to customize the name of the group from a constant or variable

Seems like we can go a long way with 1 & 2 for now; 3 seems like something to tackle once this is working in production well.

Reply:
By namespace makes the most sense with the limits imposed by CW on log puts.  Grouping all app logs into a single stream will result in loss of logs as we'll exceed quota on clusters heavily utilizing logs.

Reply:
I contend we should give user's fewer choices by introducing strategies that represent the intended data model; more open configuration options leads to more places to make the wrong decisions. I don't have a strong opinion of utilizing a single logroup versus multiple but general idea of strategies:

* Source  - one group (opinionatedly named) per source (e.g. openshift-infrastructure, openshift-application)
* Namespaced - one group per ns (opinionatedy named)? how to name logstreams is open question. 
* Custom - future choice if needed to allow users the config flexability

My original proposal would be a similar concept but utilizing a single logroup and alternate logstreams based on a similar naming strategy. Quota seems to be restricted to logstreams so I'm not certain what advantage is provided by separating by group.  Does it make query and fetch easier?  Does it allow setting permissions differently?

Reply:
+1 I'll add namespace and leave the other options as nice-to-have.